src/main/scala/io/qbeast/spark/delta/writer/SparkDeltaDataWriter.scala

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2021 Qbeast Analytics, S.L.
+ */
+package io.qbeast.spark.delta.writer
+
+import io.qbeast.IISeq
+import io.qbeast.core.model._
+import io.qbeast.spark.index.QbeastColumns
+import io.qbeast.spark.index.QbeastColumns.cubeColumnName
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.mapreduce.Job
+import org.apache.spark.qbeast.config.{MAX_FILE_SIZE_COMPACTION, MIN_FILE_SIZE_COMPACTION}
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.delta.actions.{AddFile, FileAction, RemoveFile}
+import org.apache.spark.sql.execution.datasources.OutputWriterFactory
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.{QbeastThreadUtils, SerializableConfiguration}
+import scala.collection.parallel.ForkJoinTaskSupport
+import scala.collection.parallel.immutable.ParVector
+
+/**
+ * Spark implementation of the DataWriter interface.
+ */
+object SparkDeltaDataWriter extends DataWriter[DataFrame, StructType, FileAction] {
+
+  override def write(
+      tableID: QTableID,
+      schema: StructType,
+      qbeastData: DataFrame,
+      tableChanges: TableChanges): IISeq[FileAction] = {
+
+    val sparkSession = qbeastData.sparkSession
+
+    val (factory: OutputWriterFactory, serConf: SerializableConfiguration) = {
+      val format = new ParquetFileFormat()
+      val job = Job.getInstance()
+      (
+        format.prepareWrite(sparkSession, job, Map.empty, schema),
+        new SerializableConfiguration(job.getConfiguration))
+    }
+
+    val qbeastColumns = QbeastColumns(qbeastData)
+
+    val blockWriter =
+      BlockWriter(
+        dataPath = tableID.id,
+        schema = schema,
+        schemaIndex = qbeastData.schema,
+        factory = factory,
+        serConf = serConf,
+        qbeastColumns = qbeastColumns,
+        tableChanges = tableChanges)
+    qbeastData
+      .repartition(col(cubeColumnName))
+      .queryExecution
+      .executedPlan
+      .execute
+      .mapPartitions(blockWriter.writeRow)
+      .collect()
+      .toIndexedSeq
+
+  }
+
+  /**
+   * Split the files belonging to a cube into groups
+   * that lays between [MIN_FILE_SIZE_COMPACTION, MAX_FILE_SIZE_COMPACTION]
+   * Based on groupFilesIntoBins from Delta Lake OptimizeTableCommand
+   * @param cubeStatuses the cubes statuses present in the index
+   * @return
+   */
+  def groupFilesToCompact(
+      cubeStatuses: IISeq[(CubeId, IISeq[QbeastBlock])]): IISeq[(CubeId, IISeq[QbeastBlock])] = {
+
+    // Check what cubes are suitable for compaction
+    val cubesToCompact = cubeStatuses
+      .map(g => (g._1, g._2.filter(_.size >= MIN_FILE_SIZE_COMPACTION)))
```

Comment:
This one is a bit cryptic to my taste, possibly because the letter `g` does not appear in `cubeStatuses`.